### PR TITLE
Full fix for duplicate EMTF tracks

### PR DESCRIPTION
**Pull request by abrinke1**
Cancel tracks in the same-sector if they share any LCTs, as is done in the firmware.  I thought this was already done in the EMTF emulator code, but either it wasn't, or was broken somehow by the 2nd-earliest LCT change.  The only commit with fixes from me is "EMTF cancel tracks by LCT, as in firmware" - the others are just synching with the head of l1t-integration-CMSSW_8_0_19.

### CODE REVIEW

File: L1Trigger/L1TMuonEndCap/plugins/L1TMuonEndCapTrackProducer.cc [hunk 1]
```diff
@@ -378,24 +378,37 @@ void L1TMuonEndCapTrackProducer::produce(edm::Event& ev,
...
+	    if (AllTracks_PreCancel[i2].winner.Rank() > rank2) {
+	      i2_dup = i2;
+	      rank2 = AllTracks_PreCancel[i2].winner.Rank();
+	    } // Find the highest-ranked duplicate track
+	  } // End if hits are identical
+	} // End loop over hits in track 2
+      } // End loop over hits in track 1
+    } // End second loop over tracks
+
+    // Only use ordering when ranks are equal.  Track order is not necessarily the same as FW. - AWB 21.09.16
```
Comment (thomreis) [inline L1Trigger/L1TMuonEndCap/plugins/L1TMuonEndCapTrackProducer.cc:409]: @abrinke1 do you know how the ordering is done in the FW? It looks as if the FW picks the lower quality track in some cases. Is this desired?

**Comment by thomreis** [comment]
A first test with 5k events shows that this fix removes the 5 duplicates that remained with the previous version. @rekovic can you make a tag with this please to produce ntuples?

**Comment by abrinke1** [comment]
Hi @thomreis,

I don't know how the ordering is done in FW.  This is one of the things I'm
hoping will match better in the updated emulator.  The FW always picks the
higher-quality track, but it may pick the track with higher "rank" earlier
on in the track-building process, which is an internal quality metric based
on number of hits and pattern straightness.  The "rank" does not map
one-to-one to uGMT "quality", so cases of lower "quality" cancelling higher
"quality" are theoretically possible.  How do you know that this happens in
FW?  The EMTF FW only outputs post-cancellation tracks in the DAQ, so I
don't know you would know that a higher-quality track got cancelled.

Cheers,
@abrinke1

**Comment by thomreis** [comment]
Hi Andrew,

the duplication removal bug revealed it. The emulator had two muons, one with a higher quality and one with a lower quality, and the FW had only the one with the lower quality. Assuming that the FW had at some point the same two muons as the emulator, then it must have cancelled the one with the higher quality.

Cheers,
Thomas

**Comment by jiafulow** [comment]
Hi Thomas,

Could you tell me in which events (run, lumi, event  number)  do you see this happened? I'll take a look at it.

Thank you,
Jia Fu

**Comment by thomreis** [comment]
Hi Jia Fu,

try these three examples:

```
Found a further match with dR=0.0 in run 279975:1094:1903140502:
    data: pt=2.5, eta=-1.99012494087, phi=1.96343994141, qual=4, muIdx=93
    emul: pt=2.5, eta=-1.99012494087, phi=1.96343994141, qual=4, muIdx=94
    previous matches:
    1:
    data: pt=2.5, eta=-1.99012494087, phi=1.96343994141, qual=4, muIdx=93
    emul: pt=2.5, eta=-1.99012494087, phi=1.96343994141, qual=8, muIdx=93
```

```
Found a further match with dR=0.0 in run 279975:1094:1904890167:
    data: pt=2.5, eta=-2.33812499046, phi=2.68336796761, qual=8, muIdx=96
    emul: pt=2.0, eta=-2.33812499046, phi=2.68336796761, qual=8, muIdx=97
    previous matches:
    1:
    data: pt=2.5, eta=-2.33812499046, phi=2.68336796761, qual=8, muIdx=96
    emul: pt=2.0, eta=-2.33812499046, phi=2.68336796761, qual=12, muIdx=96
```

```
Found a further match with dR=0.0 in run 279975:1095:1905180265:
    data: pt=4.0, eta=2.28375005722, phi=-0.731013298035, qual=8, muIdx=15
    emul: pt=3.5, eta=2.28375005722, phi=-0.731013298035, qual=8, muIdx=16
    previous matches:
    1:
    data: pt=2.5, eta=-2.19674992561, phi=1.60347604752, qual=12, muIdx=93
    emul: pt=2.5, eta=-2.19674992561, phi=1.60347604752, qual=12, muIdx=93
    2:
    data: pt=4.0, eta=2.28375005722, phi=-0.731013298035, qual=8, muIdx=15
    emul: pt=2.0, eta=2.28375005722, phi=-0.731013298035, qual=12, muIdx=15
```

After applying commit 6600d8b12f0e96fcd2d56a609c40276800a43453 these duplicates are gone.

Cheers,
Thomas

**Comment by jiafulow** [comment]
Hi Thomas,

Thanks! I will get back to you.

Jia Fu

**Comment by jiafulow** [comment]
Hi Thomas,

Are these events from the SingleMuon primary dataset? I tried to run on /SingleMuon/Run2016G-v1/RAW, but couldn't get the events. Maybe I didn't do it quite correctly. But I wanted to make sure that I'm processing the right dataset.

Thanks,
Jia Fu

**Comment by thomreis** [comment]
Hi Jia Fu,

the events are in this file of the ParkingZeroBias0 dataset:
/store/data/Run2016G/ParkingZeroBias0/RAW/v1/000/279/975/00000/004CF9AC-CA73-E611-B942-02163E011F66.root

Sorry, I should have mentioned it.

Cheers,
Thomas

**Comment by rekovic** [comment]
The last commit (6600d8b) of this PR has been added to the branch l1t-integration-CMSSW_8_0_19
and tagged as l**1t-integration-v86.4.**  
Please make ntuples to check for agreement in muon trigger rates between the data/emul.

**Comment by thomreis** [comment]
With l1t-integration-v86.4 I get a good but still not perfect data vs. emulator agreement for the efficiency.

https://treis.web.cern.ch/treis/cms/l1_muon_upgrade/tagAndProbe/2016_279975_singleMuon_i86p4/dataEmul_best_l1_muon_qualMin12_ptmin18_dr0.5_matched_probe_absEtaMin0_absEtaMax2.4_ptmin24.eta_over_probe_absEtaMin0_absEtaMax2.4_ptmin24.eta.png

The largest contribution to the disagreement comes from the region around phi = 1, where the efficiency is lower.

https://treis.web.cern.ch/treis/cms/l1_muon_upgrade/tagAndProbe/2016_279975_singleMuon_i86p4/dataEmul_best_l1_muon_qualMin12_ptmin18_dr0.5_matched_probe_absEtaMin1.24_absEtaMax2.4_ptmin24.phi_over_probe_absEtaMin1.24_absEtaMax2.4_ptmin24.phi.png

I also still see some muons that have duplicates with exactly the same eta and phi variables. The fraction is much smaller now. In 278074 matched muons I found the duplicates listed below. Some seem to be also duplicates that the FW did not cancel. These would be the ones where more than one further match are found for the same event number.

```
Found a further match with dR=0.0 in event 279975:121:112128610:
    data: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
    emul: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=64
    previous matches:
    1:
    data: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
    emul: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
Found a further match with dR=0.0 in event 279975:121:112128610:
    data: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=64
    emul: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
    previous matches:
    1:
    data: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
    emul: pt=4.0, eta=0.0652500018477, phi=-1.72364127636, qual=12, muIdx=63
Found a further match with dR=0.0 in event 279975:708:1171385781:
    data: pt=4.0, eta=-2.21849989891, phi=-0.447405308485, qual=12, muIdx=106
    emul: pt=4.5, eta=-2.21849989891, phi=-0.447405308485, qual=12, muIdx=105
    previous matches:
    1:
    data: pt=10.0, eta=-2.21849989891, phi=-0.447405308485, qual=8, muIdx=105
    emul: pt=4.5, eta=-2.21849989891, phi=-0.447405308485, qual=12, muIdx=105
Found a further match with dR=0.0 in event 279975:735:1217956693:
    data: pt=1.5, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=10
    emul: pt=1.5, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=10
    previous matches:
    1:
    data: pt=1.5, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=10
    emul: pt=3.0, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=9
Found a further match with dR=0.0 in event 279975:735:1217956693:
    data: pt=5.0, eta=2.22937488556, phi=-2.19268536568, qual=4, muIdx=9
    emul: pt=3.0, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=9
    previous matches:
    1:
    data: pt=1.5, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=10
    emul: pt=3.0, eta=2.22937488556, phi=-2.19268536568, qual=12, muIdx=9
Found a further match with dR=0.0 in event 279975:878:1490970691:
    data: pt=10.5, eta=2.32725000381, phi=-1.49457335472, qual=4, muIdx=12
    emul: pt=10.5, eta=2.32725000381, phi=-1.49457335472, qual=4, muIdx=12
    previous matches:
    1:
    data: pt=14.5, eta=2.32725000381, phi=-1.49457335472, qual=4, muIdx=13
    emul: pt=10.5, eta=2.32725000381, phi=-1.49457335472, qual=4, muIdx=12
Found a further match with dR=0.0 in event 279975:904:1540171675:
    data: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
    emul: pt=2.0, eta=-2.15324997902, phi=1.64710795879, qual=4, muIdx=94
    previous matches:
    1:
    data: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
    emul: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
Found a further match with dR=0.0 in event 279975:904:1540171675:
    data: pt=2.0, eta=-2.15324997902, phi=1.64710795879, qual=4, muIdx=94
    emul: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
    previous matches:
    1:
    data: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
    emul: pt=9.5, eta=-2.15324997902, phi=1.64710795879, qual=8, muIdx=93
Found a further match with dR=0.0 in event 279975:944:1612469475:
    data: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
    emul: pt=6.5, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=52
    previous matches:
    1:
    data: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
    emul: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
Found a further match with dR=0.0 in event 279975:944:1612469475:
    data: pt=6.5, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=52
    emul: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
    previous matches:
    1:
    data: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
    emul: pt=7.0, eta=-0.511125028133, phi=2.44339203835, qual=12, muIdx=51
Found a further match with dR=0.0 in event 279975:183:220564382:
    data: pt=16.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=94
    emul: pt=16.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=94
    previous matches:
    1:
    data: pt=16.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=94
    emul: pt=21.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=93
Found a further match with dR=0.0 in event 279975:183:220564382:
    data: pt=6.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=93
    emul: pt=21.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=93
    previous matches:
    1:
    data: pt=16.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=94
    emul: pt=21.5, eta=-1.72912502289, phi=1.30895996094, qual=4, muIdx=93
Found a further match with dR=0.0 in event 279975:232:309617008:
    data: pt=2.5, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
    emul: pt=3.5, eta=2.02274990082, phi=-0.0983493104577, qual=8, muIdx=15
    previous matches:
    1:
    data: pt=2.5, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
    emul: pt=3.0, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
Found a further match with dR=0.0 in event 279975:232:309617008:
    data: pt=3.0, eta=2.02274990082, phi=-0.0983493104577, qual=8, muIdx=15
    emul: pt=3.0, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
    previous matches:
    1:
    data: pt=2.5, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
    emul: pt=3.0, eta=2.02274990082, phi=-0.0983493104577, qual=12, muIdx=16
Found a further match with dR=0.0 in event 279975:436:679213503:
    data: pt=9.0, eta=-2.08800005913, phi=-0.458313316107, qual=8, muIdx=105
    emul: pt=21.0, eta=-2.08800005913, phi=-0.458313316107, qual=8, muIdx=105
    previous matches:
    1:
    data: pt=21.0, eta=-2.08800005913, phi=-0.458313316107, qual=8, muIdx=106
    emul: pt=21.0, eta=-2.08800005913, phi=-0.458313316107, qual=8, muIdx=105
Found a further match with dR=0.0 in event 279975:439:683988244:
    data: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
    emul: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=58
    previous matches:
    1:
    data: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
    emul: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
Found a further match with dR=0.0 in event 279975:439:683988244:
    data: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=58
    emul: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
    previous matches:
    1:
    data: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
    emul: pt=4.0, eta=-0.837374985218, phi=-2.58537340164, qual=12, muIdx=57
Found a further match with dR=0.0 in event 279975:476:747363221:
    data: pt=7.5, eta=1.76174998283, phi=0.850823998451, qual=8, muIdx=0
    emul: pt=4.5, eta=1.76174998283, phi=0.850823998451, qual=4, muIdx=1
    previous matches:
    1:
    data: pt=7.5, eta=1.76174998283, phi=0.850823998451, qual=8, muIdx=0
    emul: pt=7.0, eta=1.76174998283, phi=0.850823998451, qual=8, muIdx=0
Found a further match with dR=0.0 in event 279975:515:819414405:
    data: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
    emul: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=58
    previous matches:
    1:
    data: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
    emul: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
Found a further match with dR=0.0 in event 279975:515:819414405:
    data: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=58
    emul: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
    previous matches:
    1:
    data: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
    emul: pt=3.5, eta=0.717750012875, phi=-2.34539723396, qual=12, muIdx=57
Found a further match with dR=0.0 in event 279975:541:871438451:
    data: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=39
    emul: pt=4.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=36
    previous matches:
    1:
    data: pt=5.0, eta=2.38162493706, phi=2.63973593712, qual=12, muIdx=6
    emul: pt=5.0, eta=2.38162493706, phi=2.63973593712, qual=12, muIdx=6
    2:
    data: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=39
    emul: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=40
Found a further match with dR=0.0 in event 279975:541:871438451:
    data: pt=4.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=36
    emul: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=40
    previous matches:
    1:
    data: pt=5.0, eta=2.38162493706, phi=2.63973593712, qual=12, muIdx=6
    emul: pt=5.0, eta=2.38162493706, phi=2.63973593712, qual=12, muIdx=6
    2:
    data: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=39
    emul: pt=6.5, eta=-0.467624992132, phi=0.305424004793, qual=12, muIdx=40
```

Cheers,
Thomas

**Comment by thomreis** [comment]
Looking at the muon index, some of the duplicates seem to be coming from BMTF (36 <= muIdx <= 71).

**Comment by abrinke1** [comment]
Hi Thomas,

In some cases, two different tracks formed from different LCTs may have the
same uGMT eta and phi values.  This happens in both firmware and the
emulator, and is the proper behavior.  As long as the DoubleMu rates /
efficiencies agree, the emulator is behaving properly.

We have a permanent EMTF inefficiency in the negative endcap from 0.44 <
phi < 0.61 due to a dead CSC chamber (ME-1/2/4).  There was additional
inefficiency caused by issues in ME-4 sector 1 (0.26 < phi < 1.31) during
run 279975.  Ideally the emulator would handle this the same way as the
firmware, but it seems to have caused discrepancies at the few % level.

Best,
Andrew

**Comment by rekovic** [comment]
Again, code here relevant to EMTF Fix is merged in l1t-integration-CMSSW_8_0_19 as taged as v86.4.
Will close this now.

**Pull request closed without merging**